### PR TITLE
Download fossa cli if missing

### PR DESCRIPTION
**Pull request by spatten**
# Overview

[Delivers part 1 of ANE-834](https://fossa.atlassian.net/browse/ANE-834).

This PR checks to see if the fossa-cli (`fossa.exe` on Windows, `fossa` on all other platforms) exists in either the config directory or in the path.

If it does, then it returns the path that should be used to invoke the CLI.

If it does not exist, then it downloads and unzip the CLI into the config directory and returns the path to the newly unzipped file.

The config directory is the directory that contains the config file.

## Acceptance criteria

- The CLI should always exist when you run broker
- The CLI should not be downloaded if it already exists in the config directory or in your path
- We should default to using fossa from the config directory if it exists in the config directory and in the path.
- The downloaded CLI should be executable and run properly

### Test that we use fossa from your path if it exists and there is no fossa in `~/.config/fossa/broker`

Test with `fossa` in your path and no fossa in `~/.config/fossa/broker`. It should tell you that it is using fossa with a path of `fossa`:

```
which fossa
/usr/local/bin/fossa

rm -f ~/.config/fossa/broker/fossa
rm -f ~/.config/fossa/broker/debugging/trace/broker.trace
cargo run -- run

 2023-03-20T18:58:41.100717Z  INFO broker: fossa path: "fossa"
```

This should just run and not download anything

```
 ls -l ~/.config/fossa/broker/fossa 
ls: /Users/scott/.config/fossa/broker/fossa: No such file or directory
```

```
rg download_from_github ~/.config/fossa/broker/debugging/trace/broker.trace
<no results>
```

### Test that we download if there is no fossa in your path and no fossa in  `~/.config/fossa/broker`

Test with no fossa in your path and no fossa in  `~/.config/fossa/broker`. It should download fossa and put it in `~/.config/fossa/broker/fossa`, and then tell you that it is running fossa from that path.

```
mv /usr/local/bin/fossa /usr/local/bin/fossa-bak
which fossa
fossa not found
rm -f ~/.config/fossa/broker/fossa
rm -f ~/.config/fossa/broker/debugging/trace/broker.trace
cargo run -- run

  2023-03-20T19:02:32.416135Z  INFO broker: fossa path: "/Users/scott/.config/fossa/broker/fossa"

```

Test that the downloaded `fossa` works:

```
~/.config/fossa/broker/fossa --version
fossa-cli version 3.7.2 (revision 49a37c0147dc compiled with ghc-9.0)
```

###  Test that If a version already exists in `~/.config/fossa/broker/fossa`we do not download it again

Test that if a version already exists in `~/.config/fossa/broker/fossa` that we just use it and do not download it again.

```
rm -f ~/.config/fossa/broker/debugging/trace/broker.trace
ll ~/.config/fossa/broker/fossa
-rwxr-x---  1 scott  staff    72M Mar 20 12:02 /Users/scott/.config/fossa/broker/fossa

cargo run -- run

broker: fossa path: "/Users/scott/.config/fossa/broker/fossa"

ll ~/.config/fossa/broker/fossa
-rwxr-x---  1 scott  staff    72M Mar 20 12:02 /Users/scott/.config/fossa/broker/fossa

rg download_from_github ~/.config/fossa/broker/debugging/trace/broker.trace
<no results>
```

### Test that we default to using the fossa in `~/.config/fossa/broker/fossa

Put `fossa` back in your path and confirm that we default to using the fossa in `~/.config/fossa/broker/fossa`.

```
mv /usr/local/bin/fossa-bak /usr/local/bin/fossa
cargo run -- run

INFO broker: fossa path: "/Users/scott/.config/fossa/broker/fossa"
```

## Risks

We have decided to implement tests later as they were difficult to write with the code as-is and Jess is working on some fixes to make this easier.

## Checklist

- [ ] I added tests for this PR's change (or explained in the PR description why tests don't make sense).

### CODE REVIEW

File: src/download_fossa_cli.rs [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+            return Err(Error::Extracting)
+                .into_report()
+                .describe("finding fossa executable in downloaded fossa release");
+        }
+    };
+
+    let mut final_file = fs::OpenOptions::new()
+        .create(true)
+        .write(true)
+        .mode(0o770)
```
Comment (jssblck) [inline src/download_fossa_cli.rs:176]: `.mode` [is implemented on](https://doc.rust-lang.org/std/os/unix/fs/trait.OpenOptionsExt.html#tymethod.mode) `std::os::unix::fs::OpenOptionsExt` and won't work on Windows (which has no concept of unix permission modes).
Do you really need to set this to be non-default permissions?

**Review by jssblck — CHANGES_REQUESTED** [review]
Really good first shot! 

General theme here of:
- Use `.describe` and `.help` for _user facing_ guidance.
- For debugging info, either put it in the trace or put it in the actual error message.
- Keep error messages short and sweet.

File: src/download_fossa_cli.rs (hunk 1 above)
Comment (jssblck) [reply src/download_fossa_cli.rs:176]: In Windows, whether it is executable is just determined by file extension. You probably just need to do this in three steps:

1. Download the zip
2. Extract the zip to a temp location
3. Use a conditionally compiled function that moves the binary from the temp location to its destination, and on unix sets it executable (make sure to `use std::os::unix::fs::OpenOptionsExt;` inside the function, not at the top of the file).

File: src/download_fossa_cli.rs [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+        .next()
+        .ok_or(Error::FindingVersion)
+        .into_report()
+        .describe_lazy(|| format!("Parsing fossa-cli version from path {}", path))?;
+
+    if !tag.starts_with('v') {
+        return Err(Error::FindingVersion)
+            .into_report()
+            .describe_lazy(|| format!(r#"Expected tag to start with v, but found "{tag}""#))?;
+    }
```
[the remark this answers is not in the document]
Comment (spatten) [reply src/download_fossa_cli.rs]: I wasn't able to change from `.ok_or().into_report()` to `.context()`. 

I think I need to start with a fresh error here, as we're converting from an `Option` returned by `next()`, and I want to return an Err if that's `None`. Let me know if I'm missing something, though!

[the remark this answers is not in the document]
Comment (jssblck) [reply src/download_fossa_cli.rs]: Oh, no, you're right, I missed that. Good catch!

File: src/download_fossa_cli.rs [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+    let mut final_file = fs::OpenOptions::new()
+        .create(true)
+        .write(true)
+        .open(final_path)
+        .into_report()
+        .change_context(Error::FinalCopy(final_path_string.clone()))?;
+    copy(&mut zip_file, &mut final_file)
+        .into_report()
+        .change_context(Error::FinalCopy(final_path_string))?;
+    Ok(())
```
Comment (jssblck) [inline src/download_fossa_cli.rs:226]: [nit] prefer the `lazy` variant to defer this clone to if an error actually appears.

```suggestion
    let final_path_string = final_path.to_str().unwrap_or("").to_string();
    let mut final_file = fs::OpenOptions::new()
        .create(true)
        .write(true)
        .open(final_path)
        .into_report()
        .change_context_lazy(|| Error::FinalCopy(final_path_string.clone()))?;
    copy(&mut zip_file, &mut final_file)
        .into_report()
        .change_context_lazy(|| Error::FinalCopy(final_path_string))?;
    Ok(())
```

**Review by jssblck — APPROVED** [review]

**Pull request merged**